design/design-doc-observe-only-resources.md

```diff
@@ -0,0 +1,703 @@
+the external resource.
+
+```yaml
+apiVersion: ec2.aws.crossplane.io/v1beta1
+kind: VPC
+metadata:
+  annotations:
+    crossplane.io/external-name: vpc-12345678
+  name: observe-vpc
+spec:
+  managementPolicy: ObserveOnly
+  forProvider:
+    region: us-east-1
+status:
+  atProvider:
+    cidrBlock: 172.16.0.0/16
+    enableDnsHostNames: false
+    enableDnsSupport: true
+    instanceTenancy: default
+    region: us-east-1
+    tags:
+      - key: managed-by
+        value: terraform
+  conditions:
+  - lastTransitionTime: "2023-01-26T14:30:19Z"
+    reason: ReconcileSuccess
+    status: "True"
+    type: Synced
+```
+
+Please note, the `status.atProvider` will be populated with the full state of
+the resource no matter what the `managementPolicy` is. This will also help
+identify any drifts between the actual state and the desired state of the
+resource for policies other than `ObserveOnly`.
+
+**Late-initialization** of the `spec.forProvider` is an exceptional case that
+worth special consideration. We will not do late-initialization when the policy
+is `ObserveOnly`, since the primary purpose of it is getting existing defaults
+from the cloud provider and using them to represent the full desired state of
+the resource under `spec.forProvider` as Managed Resource being the source of
+truth. With `ObserveOnly` policy however, this is not the case, and it would be
+misleading if resource spec changes after the late-initialization.
+
+### Implementation
+
+Crossplane providers already manage external resources by implementing the
+Crossplane runtime's `ExternalClient` interface, which includes the four methods
+listed below.
+
+```go
+type ExternalClient interface {
+	Observe(ctx context.Context, mg resource.Managed) (ExternalObservation, error)
+	Create(ctx context.Context, mg resource.Managed) (ExternalCreation, error)
+	Update(ctx context.Context, mg resource.Managed) (ExternalUpdate, error)
+	Delete(ctx context.Context, mg resource.Managed) error
+}
+```
+
+We will leverage the fact that we have an already implemented Observe method for
+all managed resources by calling only it when the Management Policy is set to
+`ObserveOnly`. This will require minor modifications in the Managed Reconciler
+code (in the Crossplane Runtime) that will return early in the reconcile loop
+and prevent invocation of the other methods, namely, Create, Update and Delete.
+These modifications will implement the following logic at a high level:
+
+Right after the `Observe` method invocation, if `ObserveOnly`:
+
+- Return error if the resource does not exist.
+- Publish connection details.
+- Ignore late-initialization result and never call `client.Update` method to
+  update the resource spec.
+- Report success and return early.
+
+We will also need the following changes per resource:
+
+- Update the API schema to have all the fields under `spec.forProvider` under
+  `status.atProvider` as well.
```

Comment:
Would we populate these fields if the feature flag was off?

Reply:
This is tricky since we cannot control schema changes with feature flags. 
With upjet, I know that populating them is much easier than not since we are simply unmarshaling terraform fields back to crd fields. Basically, we don't need any changes in the existing logic there.
This might be the opposite for native providers. So, I would prefer to leave this open and up to implementation.